File: src/builder.java
public class Director {

    public void buildBugatti(CarBuilder builder) {
        builder.brand("Bugatti")
            .color("Blue")
            .nbrDoors(2)
            .engine("8L")
            .height(115);
    }

    public void buildLambo(CarBuilder builder) {
        builder.brand("Lamborghini")
            .model("Aventador")
            .color("Yellow")
            .nbrDoors(2)
            .height(115);
    }

}

public class CarBuilder {
    private int id;
    private String brand;
    private String model;
    private String color;

    public CarBuilder id(int id) {
        this.id = id;
        return this;
    }

    public CarBuilder brand(String brand) {
        this.brand = brand;
        return this;
    }

    public CarBuilder model(String model) {
        this.model = model;
        return this;
    }

    public CarBuilder color(String color) {
        this.color = color;
        return this;
    }

    public Car build() {
        return new Car(id, brand, model, color);
    }
}

public class Car {
    private final int id;
    private final String brand;
    private final String model;
    private final String color;

    Car(int id, String brand, String model, String color) {
        this.id = id;
        this.brand = brand;
        this.model = model;
        this.color = color;
    }
}

CarBuilder builder = new CarBuilder();
builder.id(2122)
    .brand("Bugatti")
    .model("Chiron")
    .color("Blue");
Car car = builder.build();

Director director = new Director();
CarBuilder builder = new CarBuilder();
director.buildBugatti(builder);
Car car = builder.build();

File: src/builder.ts
interface ICar {
  id?: number;
  brand?: string;
  model?: string;
  color?: string;
  engine?: string;
  qtyDoors?: number;
  height?: number;
}

class Car {
  constructor(readonly car: ICar) {}
}

class CarBuilder {
  private car: ICar = {};

  set<K extends keyof ICar>(key: K, value: ICar[K]): CarBuilder {
    this.car[key] = value;
    return this;
  }

  build(): Car {
    return new Car(this.car);
  }
}

const builder = new CarBuilder();
builder.set("id", 2122).set("brand", "Bugatti").set("model", "Chiron").set("color", "Blue");
const car = builder.build();

console.log({ car });

class Director {
  buildBugatti(builder: CarBuilder): void {
    builder.set("brand", "Bugatti").set("color", "Blue").set("qtyDoors", 2).set("engine", "8L").set("height", 115);
  }

  buildLambo(builder: CarBuilder): void {
    builder
      .set("brand", "Lamborghini")
      .set("model", "Aventador")
      .set("color", "Yellow")
      .set("qtyDoors", 2)
      .set("height", 115);
  }
}

const director = new Director();
const builderFromDirector = new CarBuilder();
director.buildBugatti(builderFromDirector);
const carFromBuilderDirector = builderFromDirector.build();

console.log({ carFromBuilderDirector });
